# FIT5196 Assessment 1: Text Processing  TASK -2
#### Student Name: Vipul Jad
#### Student ID: 29520762
##### Date:02-sep-2018
###### environment: Python 2.7.11 and Jupyter notebook Libraries used: 
###### -nltk - natural language toolkit (tokenizer, lemmatizer, stopwords, collocations and probabilities)
###### -re (for regular expression, included in Anaconda Python 2.7) 

# Creating a list of resumes and deleting Repetition

- in this cell we first imported the required libraries
- the given resumes were copied into a list
- the repetetion of resumes were removed from the list

In [1]:
import re
from nltk.stem import PorterStemmer as ps #importing porter stemmer library
from nltk.tokenize import RegexpTokenizer  #importing regex tokenizer
vocab=[]   #empty list
c=0
voc=[]
se=[535,496,467,621,735,701,191,463,629,645,230,359,272,771,606,658,865,16,777,769,129,678,181,218,137,839,851,671,377,723,825,94,263,636,746,504,456,420,651,754,421,542,86,502,727,150,856,8,854,247,207,546,
441,578,582,215,522,591,614,69,159,499,282,777,493,165,297,639,309,47,861,424,397,453,400,65,454,78,669,232,86,542,625,327,644,110,628,446,434,501,622,677,300,460,320,429,20,206,113,541,470,48,588,296,5,448,317,700,231,442,297,794,523,275,430,694,730,199,
626,478,700,387,643,724,578,284,286,394,536,607,725,459,767,831,641,289,378,471,511,335,314,705,464,184,458,602,301,266,682,805,514,545,782,313,458,529,624,639,54,583,595,579,498,356,598,47,341,537,440,696,794,175,375,388,607,501,530,819,413,636,624,371,325,
175,326,198,755,222,24,373,90,561,776,809,379,25,134,543,471,458,368,122,428,59,193,11,766,686,537,327,491,298,494,165,15,665,631,64,856,357,782,331,696,694,281,590,597,717,508,592,811,745,67,302,69,373,637,147,578,276,441,159,56,97,497,452,651,193,648,
     190]  #putting the given no. of resumes into a list
lis=[]
for i in se:
       if i not in lis:
            lis.append(i)
print(len(lis))  #removing repetitions          


219


# Reading Resumes and Sentence Segmentation

- in this cell the resume list was used to access the resumes using a for loop
- all the resumes were appended into a list
- using nltk's sent_tokenize library sentences were segmented
- all the sentences of the resume were saved into a list "sentences"


In [2]:
import nltk
nltk.download('punkt')
nl=[]
resumes=""
for i in range (0,len(lis)-1):
    string="resume_txt/resume_("+str(lis[i])+").txt"
    #print(string)
    db= open(string,'r',errors='ignore',encoding="utf-8")
    x=db.read()
    nl.append(x)
for resume in nl:
    resumes+=resume
    
#print(nl) 
print(resumes)
##sentence segmentation
from nltk.tokenize import sent_tokenize
sentences=sent_tokenize(resumes)
#print(sentences)
    

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\vipul\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Sandy K.S. Tong 

     Mobile: +852 6998 9253 
Email: sand2126sen@gmail.com 
 Address: 2A Shing Loong Court, 13 Dragon Terrace, Hong Kong 

 
 
 

Dear Hiring Manager,  

 

I  am  writing  to  you  in  regards  to  the  Division  Assistant  role posted  on  eFinancial  Careers  for  the 

China Merchants Securities International Co., Ltd. I am confident that my outgoing personality, keen attention to 

details, willingness to learn and passion for excellence will enhance the quality and culture of your company. 

 
W ith  over  6  years  of  well-rounded  secretarial  and  administrative  experience  in  supporting  senior management, 

expatriate  staffs  and  high  net  worth  clients,  I  have  demonstrated  capacity  to  develop  and  maintain 

comprehensive  administrative  processes  that  improve  the  efficiency  of  day-to-day  operations.  I  have  excelled 

my  multitasking  and  prioritizing  skills  under  pressure  through  working  in  a  fast-paced  environment  with

# Word Tokenization using Regex

- in this cell using the RegexpTokenizer we tokenize the words
- given regex was used \w+(?:[-']\w+)?
- by using the set we remove the repetetion in the list


In [3]:
from nltk.tokenize import RegexpTokenizer 
tokenize = RegexpTokenizer(r"\w+(?:[-']\w+)?")
unigram = tokenize.tokenize(resumes)
#print (unigram)
voc=list(set(unigram))
print(voc)


['Legacy', 'Sparks', 'exceptiona', 'CACS', 'determines', 'performance', 'SME-Wealth', 'Paladyne', 'DELOITTE', 'facturing', 'Changchun', 'Applicability', 'calculations', 'ar', 'Ensured', 'over-the', 'affairs', 'custody', 'reversal', 'REREFENCES', 'GBP', 'Volleyball', 'santu22', 'web', 'municipality', 'Distinction', 'Borsa', 'Storrs', 'UX', '2010', 'Cohort', 'timetables', 'usd', 'post-acquisition', 'secretarial', 'promising', 'Mareva', 'Further', '联合', 'Board', 'land-use', 'cherishhk', 'develop', 'ng', 'feedbackAssisted', 'entrusted', 'USD500m', 'domain', 'carried', 'professionalism', 'decision-making', 'miscellaneous', '91596778', 'FOONG', 'SCIENCE', 'utisling', 'generated', 'comment', 'SECURITIES', 'ang88', 'residence', 'fan', 'transferring', 'England', '5173', 'founded', 'ies', '2C', 'Financia', 'define', 'Sub-committee', 're-group', 'TAMMIE', '½', 'Adventure', 'Registers', 'coordination', 'Advice', 'insights', 'Sale', 'sufficiently', 'lient', 'Jun2013', '1A', 'Believes', 'Schedules',

# Selecting meaningful Bigrams

- in this cell we import MWETokenizer
- using this library we create bigram tokens

In [4]:
from nltk.tokenize import MWETokenizer
bigrams = MWETokenizer(voc)
b_t = bigrams.tokenize(unigram)
#print (b_t)

- find meaningful bigrams
- removing numbered words form the tokenlist
- removing words with length less than 3 to find meaningful words
- appending the first 200 bigrams into a list

In [5]:

#first 200 meaningful bigrams
bigram_list=[]
bi_2 = []
from nltk import bigrams
bgm=list(bigrams(b_t))
#print(bgm)
for i in range (0,len(bgm)-1):
    if not bgm[i][0].isdigit()  and not bgm[i][1].isdigit():
        #print(bgm[i][0])
        if len(bgm[i][0])>3 and len(bgm[i][1])>3:
            bigram_list.append(bgm[i])
for i in range(0,200):
    bi_2.append(bigram_list[i])
print(bi_2)    

[('Tong', 'Mobile'), ('Email', 'sand2126sen'), ('sand2126sen', 'gmail'), ('Shing', 'Loong'), ('Loong', 'Court'), ('Dragon', 'Terrace'), ('Terrace', 'Hong'), ('Hong', 'Kong'), ('Kong', 'Dear'), ('Dear', 'Hiring'), ('Hiring', 'Manager'), ('Division', 'Assistant'), ('Assistant', 'role'), ('role', 'posted'), ('eFinancial', 'Careers'), ('China', 'Merchants'), ('Merchants', 'Securities'), ('Securities', 'International'), ('confident', 'that'), ('outgoing', 'personality'), ('personality', 'keen'), ('keen', 'attention'), ('details', 'willingness'), ('excellence', 'will'), ('will', 'enhance'), ('your', 'company'), ('well-rounded', 'secretarial'), ('administrative', 'experience'), ('supporting', 'senior'), ('senior', 'management'), ('management', 'expatriate'), ('expatriate', 'staffs'), ('worth', 'clients'), ('have', 'demonstrated'), ('demonstrated', 'capacity'), ('maintain', 'comprehensive'), ('comprehensive', 'administrative'), ('administrative', 'processes'), ('processes', 'that'), ('that', '

# Lowercasing the letters in the sentence.

- check if the first letter in statement is uppercase
- if the second letter is uppercase then do nothing because it can be an abbrivation
- convert the first letter to lower case

In [6]:
#lowercasing letters
lower=[]
for each in sentences:
    if each[0].isupper():
        if each[1].isupper():
            continue
        else:
            #print(each[0])
            lower.append(each.replace(each[0],each[0].casefold()))
print(lower)        
        

['sandy K.s.', 'tong \n\n     Mobile: +852 6998 9253 \nEmail: sand2126sen@gmail.com \n Address: 2A Shing Loong Court, 13 Dragon terrace, Hong Kong \n\n \n \n \n\nDear Hiring Manager,  \n\n \n\nI  am  writing  to  you  in  regards  to  the  Division  Assistant  role posted  on  eFinancial  Careers  for  the \n\nChina Merchants Securities International Co., Ltd.', 'i am confident that my outgoing personality, keen attention to \n\ndetails, willingness to learn and passion for excellence will enhance the quality and culture of your company.', 'w ith  over  6  years  of  well-rounded  secretarial  and  administrative  experience  in  supporting  senior management, \n\nexpatriate  staffs  and  high  net  worth  clients,  I  have  demonstrated  capacity  to  develop  and  maintain \n\ncomprehensive  administrative  processes  that  improve  the  efficiency  of  day-to-day  operations.', 'i  have  excelled \n\nmy  multitasking  and  prioritizing  skills  under  pressure  through  working  in 

# refining wordlength , removing words with less than 3 characters

- a regex has been made to check wether the token is alphabet only and length is set from 3 to 45
- this regex will remove all unwanted numbers and words will have length from 3 to 45 as 45 is the largest recorded word in dictionary 
- after refining the tokens append them into a new list

In [7]:
#refining string removing <3 and unwanted numbers
ns=[]
reg=''
reg=re.compile(r'^[a-zA-Z]{3,45}$',re.DOTALL)
for m in voc:
    #print(m)
    if re.search(reg,m):
        ns.append(m)
#
ref_voc=ns
print(ref_voc)
       
    

['Legacy', 'Sparks', 'exceptiona', 'CACS', 'determines', 'performance', 'Paladyne', 'DELOITTE', 'facturing', 'Changchun', 'Applicability', 'calculations', 'Ensured', 'affairs', 'custody', 'reversal', 'REREFENCES', 'GBP', 'Volleyball', 'web', 'municipality', 'Distinction', 'Borsa', 'Storrs', 'Cohort', 'timetables', 'usd', 'secretarial', 'promising', 'Mareva', 'Further', 'Board', 'cherishhk', 'develop', 'feedbackAssisted', 'entrusted', 'domain', 'carried', 'professionalism', 'miscellaneous', 'FOONG', 'SCIENCE', 'utisling', 'generated', 'comment', 'SECURITIES', 'residence', 'fan', 'transferring', 'England', 'founded', 'ies', 'Financia', 'define', 'TAMMIE', 'Adventure', 'Registers', 'coordination', 'Advice', 'insights', 'Sale', 'sufficiently', 'lient', 'Believes', 'Schedules', 'houses', 'Republic', 'achieving', 'swinburne', 'proceedings', 'exit', 'average', 'Sud', 'summary', 'entities', 'manages', 'launches', 'about', 'Continuously', 'Efficiency', 'teammates', 'Ensure', 'Sport', 'destress'

# Removing Stopwords

- stopwords english library is used
- with for loop check if the token exists in stopwords if not then append into the list

In [8]:
from nltk.corpus import stopwords
swl = stopwords.words('english')
sws = set(swl)
st_voc = [w for w in ref_voc if w not in sws]
print (st_voc)

['Legacy', 'Sparks', 'exceptiona', 'CACS', 'determines', 'performance', 'Paladyne', 'DELOITTE', 'facturing', 'Changchun', 'Applicability', 'calculations', 'Ensured', 'affairs', 'custody', 'reversal', 'REREFENCES', 'GBP', 'Volleyball', 'web', 'municipality', 'Distinction', 'Borsa', 'Storrs', 'Cohort', 'timetables', 'usd', 'secretarial', 'promising', 'Mareva', 'Further', 'Board', 'cherishhk', 'develop', 'feedbackAssisted', 'entrusted', 'domain', 'carried', 'professionalism', 'miscellaneous', 'FOONG', 'SCIENCE', 'utisling', 'generated', 'comment', 'SECURITIES', 'residence', 'fan', 'transferring', 'England', 'founded', 'ies', 'Financia', 'define', 'TAMMIE', 'Adventure', 'Registers', 'coordination', 'Advice', 'insights', 'Sale', 'sufficiently', 'lient', 'Believes', 'Schedules', 'houses', 'Republic', 'achieving', 'swinburne', 'proceedings', 'exit', 'average', 'Sud', 'summary', 'entities', 'manages', 'launches', 'Continuously', 'Efficiency', 'teammates', 'Ensure', 'Sport', 'destress', 'fulfil

# removing rare tokens from the token list

- first get a dictionary with word count of every token in the resume
- update the dictionary 

In [9]:
wf={}
for word in st_voc:
    c=0
    for res in nl:
        if word in res:
            c+=1
    if c!=0:
        wf.update({word:c})
print(wf)    
                

{'Legacy': 1, 'Sparks': 1, 'exceptiona': 8, 'CACS': 2, 'determines': 1, 'performance': 78, 'Paladyne': 1, 'DELOITTE': 4, 'facturing': 21, 'Changchun': 1, 'Applicability': 1, 'calculations': 14, 'Ensured': 10, 'affairs': 6, 'custody': 8, 'reversal': 1, 'REREFENCES': 1, 'GBP': 2, 'Volleyball': 3, 'web': 14, 'municipality': 1, 'Distinction': 27, 'Borsa': 1, 'Storrs': 1, 'Cohort': 3, 'timetables': 1, 'usd': 1, 'secretarial': 17, 'promising': 2, 'Mareva': 1, 'Further': 6, 'Board': 21, 'cherishhk': 1, 'develop': 95, 'feedbackAssisted': 1, 'entrusted': 1, 'domain': 1, 'carried': 10, 'professionalism': 1, 'miscellaneous': 2, 'FOONG': 1, 'SCIENCE': 2, 'utisling': 1, 'generated': 12, 'comment': 9, 'SECURITIES': 2, 'residence': 2, 'fan': 5, 'transferring': 1, 'England': 9, 'founded': 6, 'ies': 214, 'Financia': 132, 'define': 5, 'TAMMIE': 1, 'Adventure': 1, 'Registers': 1, 'coordination': 12, 'Advice': 1, 'insights': 8, 'Sale': 34, 'sufficiently': 1, 'lient': 178, 'Believes': 1, 'Schedules': 1, 'h

- we need to remove the words which are only in 2% of the list
- in this case 2% of 219 files is 4.3 which is rounded to 5
- if the wordcount is less than 5 remove the word

In [10]:
nd={}
ll=[]
for x,y in wf.items():
    if y>5:
        nd.update({x:y})
        ll.append(x)
#print(nd)    
print(ll)    

['exceptiona', 'performance', 'facturing', 'calculations', 'Ensured', 'affairs', 'custody', 'web', 'Distinction', 'secretarial', 'Further', 'Board', 'develop', 'carried', 'generated', 'comment', 'England', 'founded', 'ies', 'Financia', 'coordination', 'insights', 'Sale', 'lient', 'houses', 'Republic', 'exit', 'average', 'summary', 'entities', 'Ensure', 'Sport', 'quo', 'liaison', 'size', 'speaking', 'Touch', 'Brokers', 'maintain', 'prepared', 'schedule', 'ean', 'stud', 'inves', 'Busines', 'rapport', 'Group', 'His', 'instrument', 'organization', 'familiar', 'Fai', 'improve', 'CERTIFICATIONS', 'facilities', 'Raise', 'variance', 'currently', 'mainly', 'example', 'soft', 'even', 'negotiations', 'Alternat', 'bonus', 'speaker', 'pursuing', 'Assessed', 'audits', 'ixed', 'Southeast', 'Micro', 'Korea', 'UNIVERSITY', 'spend', 'User', 'Extra', 'REST', 'RSM', 'fun', 'involved', 'allow', 'academ', 'ASE', 'pricing', 'agreed', 'Transfer', 'itor', 'SCH', 'payable', 'Cheung', 'buyer', 'Employment', 'Pro

# STEMMING
- using stemming.porter 2 library
- append the stemmed tokens into a list

In [11]:
fin=[]
from stemming.porter2 import stem
fin +=list(set( [stem(word) for word in ll]))

In [12]:
print(fin)

['aud', 'Pat', 'exceptiona', 'engag', 'Her', 'ted', 'expans', 'adv', 'mediat', 'Nov', 'particular', 'inv', 'entir', 'Femal', 'Contribut', 'sen', 'execu', 'web', 'inal', 'receipt', 'care', 'metric', 'Softw', 'Busi', 'ser', 'Further', 'incent', 'Board', 'Best', 'Las', 'develop', 'dministr', 'Employe', 'sourc', 'oduc', 'achiev', 'alloc', 'comment', 'art', 'Pub', 'Englis', 'England', 'qualit', 'loba', 'ran', 'approv', 'Financia', 'point', 'inner', 'sever', 'SINGAPORE', 'identifi', 'orient', 'Sale', 'lient', 'QUALIFICATION', 'rev', 'Allianc', 'sse', 'state', 'live', 'jor', 'exit', 'scholarship', 'Oct', 'quota', 'Africa', 'previous', 'divis', 'situat', 'MIT', 'Charter', 'guidelin', 'Ensure', 'Sport', 'ACA', 'Qual', 'amount', 'quo', 'liaison', 'Interact', 'size', 'Post', 'yea', 'CHU', 'templat', 'Auto', 'Touch', 'citi', 'Energi', 'portfo', 'igh', 'maintain', 'GAAP', 'Provid', 'Franc', 'fluen', 'handl', 'WORK', 'initi', 'cumu', 'ean', 'firm', 'complet', 'Lang', 'inanci', 'Cum', 'Philipp', 'stu

# writing the files with Bigrams and Unigrams with their indexes
- open a file with write permission
- with for loop write the first 200 meaningful bigrams in it
- do same with the unigrams

In [16]:
#writing file 
dx=open("29520762_vocab1.txt","w")
bindx=1
windx=1
a=''
b=''
tokl=[]
fin.sort()
for words in fin:
    dx.write(words+":"+str(windx)+"\n")
    tokl.append((words,windx))
    windx+=1
for bis in bi_2:
    #print(bis[0]+bis[1])
    a=str(bis[0])
    b=str(bis[1])
    #print(type(a)
    dx.write("("+a+","+str(b)+") :"+str(bindx)+"\n")
    bindx+=1


dx.close()


# word count per resume with word ID
- open a file with write permissions
- with for loop in range from 0 to length of list of resumes
- write resume name of that pirticular iteration
- with for loop of tokenlist
- count each word in resume
- if count > 0 write count

In [17]:
op=open("29520762_countVec.txt","w")

In [18]:
c=0
for i in range (0,len(lis)-1):
    op.write("\n resume_("+str(lis[i])+"),\n")
    for word in tokl:
        c=nl[i].count(word[0])
        if c>0:
            op.write(str(word[1])+":"+str(c)+",")  
op.close()            